wire-ios-sync-engine/Source/UserSession/ZMUserSession.swift

```diff
@@ -624,28 +641,31 @@ extension ZMUserSession: ZMSyncStateDelegate {
         }
     }
 
-    public func didFinishQuickSync() {
+    public func didFinishQuickSync() async {
         Self.logger.trace("did finish quick sync")
-        processEvents()
+        await processEvents()
 
-        syncContext.mlsService?.performPendingJoins()
+        syncContext.performAndWait {
+            syncContext.mlsService?.performPendingJoins()
 
-        managedObjectContext.performGroupedBlock { [weak self] in
-            self?.notifyThirdPartyServices()
+            managedObjectContext.performGroupedBlock { [weak self] in
+                self?.notifyThirdPartyServices()
+            }
         }
-
-        commitPendingProposalsIfNeeded()
+        await commitPendingProposalsIfNeeded()
         fetchFeatureConfigs()
         recurringActionService.performActionsIfNeeded()
     }
 
-    func processEvents() {
+    func processEvents() async {
```

Comment:
Unsure if should make this async, didn't we discuss making the event processor an actor so that we can safely process events different queues and be sure that processing happens sequentially.